### ISSUE TITLE
Output difference from FasterTransformer

### ISSUE THREAD

**Issue by sleepwalker2017** [InternLM/lmdeploy#444]
我在对比FT 和 Lmdeploy 输出结果的一致性。我打印了 sampling 之前的 logits，发现差异比较大。
可能是内部哪些地方做过修改吗？如果算子是沿用的 FT 的，那不应该有这么大的 diff 啊？

我的输入batch 是 1，打印了每个 step 输出的logits，并找出其中的最大值和下标。

可以看到，diff 普遍有些大，红框中的内容是差异非常大了。

<img width="921" alt="image" src="https://github.com/InternLM/lmdeploy/assets/26128514/fc03e99d-8319-4cb8-8605-6ac451d85efa">

**Comment by lvhan028** [comment]
算子改了很多。
我们使用 opencompass 测过推理精度，是和算法侧对齐的

**Comment by sleepwalker2017** [comment]
> 算子改了很多。 我们使用 opencompass 测过推理精度，是和算法侧对齐的

感谢解答！能简单举个例子吗   是attention这种算子做了修改？

另外，推理精度是指类似 PPL 这样的指标吗？推理结果有 diff，但是算法指标一致，是这个意思吗？

另外一个问题，lmdeploy 在用 tokenizer 的 encode 阶段，默认add_special_tokens是 False，我看FT的 demo里使用了tokenizer的默认值(True)。这个是出于什么考虑吗？我发现这个对结果一致性影响还挺大的。
```python
    def encode(self, s: str):
        """Tokenize a prompt.

        Args:
            s (str): a prompt
        Returns:
            list[int]: token ids
        """
        add_special_tokens = False
        if s.find('<BOS>') != -1:
            s = s.replace('<BOS>', '<s>')
        if s == '<EOS>':
            s = '</s>'
        if len(s) == 0:
            add_special_tokens = True
        return self.model.encode(s, add_special_tokens=add_special_tokens)
```

**Comment by sleepwalker2017** [comment]
> 算子改了很多。 我们使用 opencompass 测过推理精度，是和算法侧对齐的

这个问题已经解决，经过排查，发现两个地方：

1. tokenizer 阶段的是add_special_token 两边不一致，导致 token_ids 有差异。
2. LayerNorm 层的norm_eps的值不一样(1e-6 vs 1e-5)，导致了计算差异。

目前数值上已经完全对得上了，感谢！
